Fixed #23004 -- Cleanse sensitive request.META keys from DEBUG Views

maxsond opened the pull request:
- Fixed ticket [#23004](https://code.djangoproject.com/ticket/23004)
- Signed and sent CLA agreement
- Includes tests and documentation

This is based almost entirely on [PR 7996](https://github.com/django/django/pull/7996). There should be two differences: 
- It works with the latest version of the master branch
- It includes the test for ExceptionReporterFilter.get_safe_request_meta that was requested at the end of that PR

audiolion commented:
Thanks for fixing what I never did. I wish that the original attribution from my commits was not lost, but I suppose I can't complain after having left that PR languishing for so long.

You are missing one feedback comment Tim made https://github.com/django/django/pull/7996/files#r99472959

charettes commented:
@audiolion the merger should take care of giving you attribution [through `Co-authored-by`](https://help.github.com/articles/creating-a-commit-with-multiple-authors/) if the final patch doesn't diverge too much from the original.

maxsond commented:
@audiolion Thanks for the catch on the missing documentation line.
 
I had some semi-informal attribution in the commit message, but I didn't know about Co-authored by. I'll rewrite the commit message to include that.

audiolion commented:
Woah TIL about `Co-authored-by` as well, that is really awesome Charettes!

maxsond commented:
Whoops, my attempt at merge-and-squashing in my own PR when I added the missing documentation change just resulted in a second commit. But if I understand correctly then whoever ends up committing this to the Django project will take care of properly squashing/attributing?

carltongibson commented on docs/howto/error-reporting.txt:
```diff
@@ -138,12 +138,18 @@ exception raised, each `traceback frame`_’s local variables, and the
+production environment (that is, where :setting:`DEBUG` is set to ``False``):
+:func:`sensitive_variables` and :func:`sensitive_post_parameters`.
+
+.. versionchanged:: 2.2
```
These need updating to 3.0

maxsond commented:
Thanks for the catch, @carltongibson. I overlooked that comment initially. I've moved those into the class and updated the documentation. I added the values as optional construction arguments with the defaults coming from constants defined on the class. The class definition looks a little clunkier but my thinking was this may be more pleasant to use when actually creating subclasses than needing to update the attributes only after construction.

Still, I'm not married to this approach and can remove the arguments (and make any other changes) if the approach in the new commits doesn't sit right with people.

ohlr commented:
I just stumbled over this as `KEY` and `PASS` haven't been filtered out of my `request.META` error report. Is there anything left to do? Or when will this be merged?

carltongibson commented:
Hi @ohlr (and @maxsond) — I hadn't noticed the updates. I've just updated the ticket to put it back in the review queue. ([Patch review checklist is here](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/submitting-patches/#patch-review-checklist) if you want to make a head-start on that. 🙂)

carltongibson commented:
(@ohlr Thanks for having a look! Check out the [Triage Workflow docs](https://docs.djangoproject.com/en/dev/internals/contributing/triaging-tickets/#triage-workflow) — There's a super power-up available to you now. Hint: sentence begins _"Daisy reviews the pull request and ..."_ Welcome aboard! 🙂)

ohlr commented on docs/howto/error-reporting.txt:
```diff
@@ -263,23 +269,43 @@ Your custom filter class needs to inherit from
+
+.. versionadded:: 3.0
+
+    * ``cleansed_substitute``: The value to replace sensitive values with.
```
```suggestion
    * ``cleansed_substitute``: The value to replace sensitive values with. Defaults to '********************'.
```

ohlr commented on docs/howto/error-reporting.txt:
```diff
@@ -263,23 +269,43 @@ Your custom filter class needs to inherit from
+.. versionadded:: 3.0
+
+    * ``cleansed_substitute``: The value to replace sensitive values with.
+    Defaults to '********************'
```
```suggestion
```

maxsond commented:
Thanks for the docs fix, @ohlr. I've committed the changes. And my bad for forgetting to take the needs_better_patch flag off the ticket.

ohlr commented:
The test `isort` check complains about:
ERROR: /tests/view_tests/tests/test_debug.py Imports are incorrectly sorted.

```
from django.views.debug import (
    CallableSettingWrapper, ExceptionReporter,
    ExceptionReporterFilter, technical_500_response,
)

from ..views import (
    custom_exception_reporter_filter_view, index_page,
    multivalue_dict_key_error, non_sensitive_view, paranoid_view,
    sensitive_args_function_caller, sensitive_kwargs_function_caller,
    sensitive_method_view, sensitive_view,
)
```
any idea why this is the case?

maxsond commented:
I'm not sure exactly how isort works, but the contribution docs said to just run it, so I ran it on the file and it bumped ExceptionReporterFilter up one line. Looks like to get the classes all on one line together, since a class and a function had been mixed on one line.

maxsond commented:
Sure! I'm excited to get this wrapped up. I'll dig into the comments you provided here after work today.

carltongibson commented on django/views/debug.py:
```diff
@@ -112,6 +72,16 @@ class ExceptionReporterFilter:
     """
 
+    DEFAULT_CLEANSED_SUBSTITUTE = '********************'
+    DEFAULT_HIDDEN_SETTINGS = re.compile('API|TOKEN|KEY|SECRET|PASS|SIGNATURE', flags=re.IGNORECASE)
```
Have these as just `CLEANSED_SUBSTITUTE` and `HIDDEN_SETTINGS`.

carltongibson commented on django/views/debug.py:
```diff
@@ -112,6 +72,16 @@ class ExceptionReporterFilter:
+    DEFAULT_CLEANSED_SUBSTITUTE = '********************'
+    DEFAULT_HIDDEN_SETTINGS = re.compile('API|TOKEN|KEY|SECRET|PASS|SIGNATURE', flags=re.IGNORECASE)
+
+    def __init__(self, cleansed_substitute=DEFAULT_CLEANSED_SUBSTITUTE, hidden_settings=DEFAULT_HIDDEN_SETTINGS):
```
Make these defaults `None`. (Then below set `self.CLEANSED_SUBSTITUTE` if `not None`.)

carltongibson commented on docs/howto/error-reporting.txt:
```diff
@@ -138,12 +138,25 @@ exception raised, each `traceback frame`_’s local variables, and the
+
+.. versionchanged:: 3.0
+
+    Django cleanses settings and :attr:.HttpRequest.META that appear sensitive,
```
I didn't mean in my previous comments to duplicate the exisiting text, rather to rephrase the section.

carltongibson commented:
Closing in favour of  #12300. Thanks @maxsond!
